storage/replica_leader_lease.go

```diff
@@ -20,100 +20,164 @@
+	startStasis := timestamp.Add(int64(replica.store.ctx.leaderLeaseActiveDuration), 0)
+	expiration := startStasis.Add(int64(replica.store.Clock().MaxOffset()), 0)
+	leaseReq := roachpb.LeaderLeaseRequest{
+		Span: roachpb.Span{
+			Key: startKey,
+		},
+		Lease: roachpb.Lease{
+			Start:       timestamp,
+			StartStasis: startStasis,
+			Expiration:  expiration,
+			Replica:     nextLeader,
+		},
+	}
+	if !replica.store.Stopper().RunAsyncTask(func() {
+		pErr := replica.executeLeaderLeaseCommand(leaseReq, timestamp, startKey)
 
 		// Send result of leader lease to all waiter channels.
-		r.mu.Lock()
-		defer r.mu.Unlock()
-		for i, llChan := range r.mu.llChans {
-			// Don't send the same pErr object twice; this can lead to
-			// races. We could clone every time but it's more efficient to
-			// send pErr itself to one of the channels (the last one; if we
-			// send it earlier the race can still happen).
-			if i == len(r.mu.llChans)-1 {
+		replica.mu.Lock()
+		defer replica.mu.Unlock()
+		for i, llChan := range p.llChans {
+			// Don't send the same pErr object twice; this can lead to races. We could
+			// clone every time but it's more efficient to send pErr itself to one of
+			// the channels (the last one; if we send it earlier the race can still
+			// happen).
+			if i == len(p.llChans)-1 {
 				llChan <- pErr
 			} else {
 				llChan <- protoutil.Clone(pErr).(*roachpb.Error) // works with `nil`
 			}
 		}
-		r.mu.llChans = r.mu.llChans[:0]
+		p.llChans = p.llChans[:0]
+		p.nextLease = roachpb.Lease{}
 	}) {
 		// We failed to start the asynchronous task. Send a blank NotLeaderError
 		// back to indicate that we have no idea who the leader might be; we've
 		// withdrawn from active duty.
-		llChan <- roachpb.NewError(r.newNotLeaderErrorLocked(nil, r.store.StoreID()))
+		llChan <- roachpb.NewError(
+			replica.newNotLeaderError(nil, replica.store.StoreID(), replica.mu.desc))
 		return llChan
 	}
-
-	r.mu.llChans = append(r.mu.llChans, llChan)
+	p.llChans = append(p.llChans, llChan)
+	p.nextLease = leaseReq.Lease
 	return llChan
 }
+
+func (r *Replica) getOrExtendLeaderLeaseLocked(timestamp roachpb.Timestamp) <-chan *roachpb.Error {
+	// Propose a Raft command to get a leader lease for this replica.
+	desc := r.mu.desc
+	_, replica := desc.FindReplica(r.store.StoreID())
+	if replica == nil {
+		llChan := make(chan *roachpb.Error, 1)
+		llChan <- roachpb.NewError(roachpb.NewRangeNotFoundError(r.RangeID))
+		return llChan
+	}
+	if r.store.IsDrainingLeadership() {
+		// We've retired from active duty.
+		llChan := make(chan *roachpb.Error, 1)
+		llChan <- roachpb.NewError(r.newNotLeaderError(nil, r.store.StoreID(), r.mu.desc))
+		return llChan
+	}
+	return r.mu.pendingLeaderLeaseRequest.InitOrJoinRequest(
+		r, *replica, timestamp, desc.StartKey.AsRawKey())
+}
+
+// executeLeaderLeaseCommand proposes a LeaderLeader and waits for its
+// completion.
+func (r *Replica) executeLeaderLeaseCommand(
```

Comment:
sorry, probably better `s/execute/send/` - `execute` sounds wrong (but `propose` doesn't quite hit the nail on the head as well).